text/0250-Composite_Samplers.md

```diff
@@ -0,0 +1,293 @@
+
+```
+S4 = Conjunction(S3, RateLimitingSampler(1000 * 60))
+```
+
+### Limitations of composite samplers in Approach One
+
+Not all samplers can participate as components of composite samplers without undesired or unexpected effects. Some samplers require that they _see_ each `Span` being created, even if the span is going to be dropped. Some samplers update the trace state or maintain internal state, and for their correct behavior it it is assumed that their sampling decisions will be honored by the tracer at the face value in all cases. A good example for this are rate limiting samplers which have to keep track of the rate of created spans and/or the rate of positive sampling decisions.
+
+The need to encode and decode the `Tracestate` multiple times affects performance of the composite samplers. This drawback is eliminated in Approach Two.
+
+## Approach Two
+
+A principle of operation for Approach Two is that `ShouldSample` is invoked only once, on the root of the tree formed by composite samplers. All the logic provided by the composition of samplers is handled by calculating the threshold values, delegating the calculation downstream as necessary.
+
+### New API
+
+To make this approach possible, all Consistent Probability Samplers which participate in the samplers composition need to implement the following API, in addition to the standard Sampler API. We will use the term _Composable Sampler_ to denote Consistent Probability Samplers which provide the new API and conform to the rules described here.
+The composite samplers in Approach Two are Composable Samplers as well.
+
+#### GetSamplingIntent
+
+This is a routine/function/method for all Composable Samplers. Its purpose is to query the sampler about the activities it would perform had it been asked to make a sampling decision for a given span, however, without constructing the actual sampling Decision.
+
+#### Required Arguments for GetSamplingIntent:
+
+The arguments are the same as for [`ShouldSample`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#shouldsample) except for the `TraceId`.
+
+- `Context` with parent `Span`.
+- Name of the `Span` to be created.
+- `SpanKind` of the `Span` to be created.
+- Initial set of `Attributes` of the `Span` to be created.
+- Collection of links that will be associated with the `Span` to be created.
+
+#### Return value:
+
+The return value is a structure (`SamplingIntent`) with the following elements:
+
+- The THRESHOLD value represented as a 14-character hexadecimal string, with value of `null` representing non-probabilistic `DROP` decision,
```

Comment:
Suggest permitting an implementation to use an unsigned representation, too.
```suggestion
- The THRESHOLD value indicating the 56-bits defined in OTEP 235, with a special value (e.g., `null`) representing the non-probabilistic `DROP` decision,
```

Reply:
(Now I see you used "lexicographical" below, which explains why you spelled out the 14-character hex-string part here. I can accept it this way, but in https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/93a78c6ad8f57c5e4ff29a5f41779af03fc85145/pkg/sampling/threshold.go#L37 I've used an unsigned representation ;)